src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs

```diff
@@ -512,41 +512,26 @@ public override long Seek(long offset, SeekOrigin loc)
         {
             EnsureNotClosed();
 
-            if (offset > MemStreamMaxLength)
-                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_StreamLength);
-
-            switch (loc)
+            return SeekCore(offset, loc switch
             {
-                case SeekOrigin.Begin:
-                    {
-                        int tempPosition = unchecked(_origin + (int)offset);
-                        if (offset < 0 || tempPosition < _origin)
-                            throw new IOException(SR.IO_SeekBeforeBegin);
-                        _position = tempPosition;
-                        break;
-                    }
-                case SeekOrigin.Current:
-                    {
-                        int tempPosition = unchecked(_position + (int)offset);
-                        if (unchecked(_position + offset) < _origin || tempPosition < _origin)
-                            throw new IOException(SR.IO_SeekBeforeBegin);
-                        _position = tempPosition;
-                        break;
-                    }
-                case SeekOrigin.End:
-                    {
-                        int tempPosition = unchecked(_length + (int)offset);
-                        if (unchecked(_length + offset) < _origin || tempPosition < _origin)
-                            throw new IOException(SR.IO_SeekBeforeBegin);
-                        _position = tempPosition;
-                        break;
-                    }
-                default:
-                    throw new ArgumentException(SR.Argument_InvalidSeekOrigin);
-            }
+                SeekOrigin.Begin => _origin,
+                SeekOrigin.Current => _position,
+                SeekOrigin.End => _length,
+                _ => throw new ArgumentException(SR.Argument_InvalidSeekOrigin)
+            });
+        }
 
-            Debug.Assert(_position >= 0, "_position >= 0");
-            return _position;
+        private long SeekCore(long offset, int loc)
+        {
+            if (offset > MemStreamMaxLength - loc)
+                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_StreamLength);
+            int tempPosition = unchecked(loc + (int)offset);
+            if (unchecked(loc + offset) < _origin || tempPosition < _origin)
```

Comment:
```suggestion
            if (tempPosition < _origin)
```

I don't see the point of the `unchecked(loc + offset) < _origin` check, furthermore, if I remove it, no test fails.

Reply:
This extra check was done previously and so I kept it.
I am not able to find why/when it was introduced as it seems to come from .NET Framework times.
I trust you on this one.

Reply:
I have removed "your" commit.